### DISCUSSION TITLE
Sort filtered results in <MultiSelect /> (or <Select /> and <Autocomplete />)

### DISCUSSION THREAD

**Discussion by ansonkao** [mantinedev/mantine#3960] [Q&A] [answered]
Is there an existing way to sort filtered results with these components?

Currently these components offer a `filter` prop with the signature `(value: string, selected: boolean, item: SelectItem) => boolean` which allows filtering but not sorting the order of the results. Sorting the order is critical for implementing fuzzy search in autocompletion.

Happy to submit a PR if no such solution exists. I see this was already requested in https://github.com/mantinedev/mantine/discussions/1330 

----

Edit: An additional huge win would be a way to highlight search terms in the results. For example, if the search query was "apple" and the results returned both "apples" and "pineapples", having the results appear as "**apple**s" and "pine**apple**s"

**Comment by rtivital** [comment]
Currently MultiSelect and other combobox components present the data based on the order that it is passed to the `data` prop. It is not planned to change these components until 7.0 release.

**Reply by ansonkao** [reply]
Thanks. Shared a workaround below in the meantime!

**Comment by ansonkao** [comment, marked as answer]
I found a workaround for anyone who wants to sort the results. Basically the solution is to update the `data` prop via `setState()` inside the `onSearchChange` callback.

Here's a pseudocode outline:

```jsx
import { useState } from 'react'
import { MultiSelect } from '@mantine/core'

export default function PosePromptBox() {
  const [selections, setSelections] = useState()
  const [selectData, setSelectData] = useState([/* ...your original data... */])
  const [searchQuery, setSearchQuery] = useState('')

  const onQuery = (query) => {
    const sortedResults = selectData.slice().sort((a, b) => {
      /* ...insert your search logic here based on the query... */
      return bScore - aScore
    })

    setSelectData(sortedResults) // <-- THIS IS THE HACK: update the `data` prop via `setState()` inside the `onSearchChange` callback
    setSearchQuery(query)
  }

  const onFilter = (value, selected, item) => {
    if (selected) return false
    if (/* ...filter your items... */) {
      return true
    }
    return false
  }

  const onChange = (value) => {
    /* ...run any business logic necessary to understand the selected result here... */
    setSelections(result)
  }

  return (
    <MultiSelect
      data={selectData}
      value={selections}
      onChange={onChange}
      searchable
      searchValue={searchQuery}
      onSearchChange={onQuery}
      filter={onFilter}
    />
  )
}

```
